Add markerscope linter

nayuta723 opened the pull request:
Fixes #148 

Implements a new `markerscope` linter that validates markers are applied in the correct scope (field vs type) and to compatible data types, ensuring proper usage of kubebuilder/controller-runtime annotations throughout Kubernetes API definitions.

**Key Features:**

- **Scope Validation**: Ensures markers are placed on the correct Go language construct
  - `FieldScope`: Can only be applied to struct fields (e.g., `optional`, `required`, `nullable`)
  - `TypeScope`: Can only be applied to type definitions (e.g., `kubebuilder:validation:items:ExactlyOneOf`)
  - `AnyScope`: Can be applied to either fields or type definitions (e.g., `Minimum`, `Pattern`)
  - **Note**: CRD-level markers (e.g., `kubebuilder:resource`, `kubebuilder:subresource`) and package-level markers (e.g., `groupName`) are out of scope for this PR

- **Type Constraint Validation**: Ensures markers are applied to compatible OpenAPI schema types
  - Numeric markers (`Minimum`, `Maximum`, `MultipleOf`) only for `integer`/`number` types
  - String markers (`Pattern`, `MinLength`, `MaxLength`) only for `string` types
  - Array markers (`MinItems`, `MaxItems`, `UniqueItems`) only for `array` types
  - Object markers (`MinProperties`, `MaxProperties`) only for `object` types (struct/map)
  - Array items markers (`items:Minimum`, `items:Pattern`, etc.) apply constraints to array element types

- **Strict Type Constraint Mode**: `strictTypeConstraint` flag enforces that markers with `AnyScope` and type constraints must be declared on type definitions rather than fields when using named types, promoting consistent marker placement

- **Automatic Fix Suggestions**: When `policy: SuggestFix` is configured, provides automatic fix suggestions for violations:
  - Move markers from incorrect scope (field ↔ type)
  - Remove markers with incompatible type constraints
  - Relocate markers to type definitions for named types

- **Configurable Rules**: Supports overriding default rules and adding custom marker validation
  - Override built-in marker rules
  - Define custom markers with scope and type constraints
  - Configure dangerous type handling (float32/float64)

- **Comprehensive Default Rules**: Includes 100+ built-in rules for:
  - All standard kubebuilder validation markers
  - k8s declarative validation markers
  - Controller-runtime markers

nayuta723 committed:
initial commit for markerscope linter

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
Add ScopeViolation

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
Add TypeViolation

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
add custom marker rule setting

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
Add custom marker rules and validation logic.

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
Enhance marker scope validation by introducing allowDangerousTypes flag and updating type constraints. Refactor schema type definitions and improve error handling for invalid markers. Add new test cases for various marker scenarios.

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
Refactor error handling in marker scope validation by introducing custom error types for invalid schema types, type constraints, and element constraints. Update validation functions to return these new error types for improved clarity and maintainability.

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
refactor(markerscope): fix linters and add test golden files

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
docs(markerscope): document strictTypeConstraint and scope exclusions

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

JoelSpeed commented on pkg/analysis/markerscope/schema.go:
```diff
@@ -0,0 +1,125 @@
+	SchemaTypeArray SchemaType = "array"
+	// SchemaTypeObject represents struct/map types.
+	SchemaTypeObject SchemaType = "object"
+)
```
Does this all need to be exported?

nayuta723 replied:
This is configured this way because it's necessary for the Configuration settings.

JoelSpeed commented on pkg/analysis/markerscope/analyzer_test.go:
```diff
@@ -0,0 +1,83 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package markerscope
```
If you move this to `markerscope_test` and use the public initilizer interface, I think that will be more in-keeping with other analyzers

JoelSpeed commented on pkg/analysis/markerscope/analyzer_test.go:
```diff
@@ -0,0 +1,83 @@
+	"golang.org/x/tools/go/analysis/analysistest"
+)
+
+func TestAnalyzerWarnOnly(t *testing.T) {
```
We normally have a test for the default configuration, what is the default? Is defaulting of the config implemented?

nayuta723 replied:
The following is the default configuration. I will add a defaultConfig function.
- Policy: Warn
- OverrideMarkers: empty (use built-in defaults)
- CustomMarkers: empty

JoelSpeed commented on pkg/analysis/markerscope/config.go:
```diff
@@ -0,0 +1,595 @@
+
+	// AnyScope indicates the marker can be placed on either fields or types.
+	AnyScope = FieldScope | TypeScope
+)
```
We don't use this kind of config for public APIs. Use a string type with PascalCase values

JoelSpeed commented on pkg/analysis/markerscope/config.go:
```diff
@@ -0,0 +1,595 @@
+	//       scope: any
+	//       typeConstraint:
+	//         allowedSchemaTypes: ["string"]
+	MarkerRules map[string]MarkerScopeRule `json:"markerRules,omitempty"`
```
Don't use a map, use a list with a name key or something to identify the marker

JoelSpeed commented on pkg/analysis/markerscope/config.go:
```diff
@@ -0,0 +1,595 @@
+	// AllowDangerousTypes specifies if dangerous types are allowed.
+	// If true, dangerous types are allowed.
+	// If false, dangerous types are not allowed.
+	AllowDangerousTypes bool `json:"allowDangerousTypes,omitempty"`
```
What's a dangerous type and why does that matter?

JoelSpeed commented on pkg/analysis/markerscope/config.go:
```diff
@@ -0,0 +1,595 @@
+// this linter's current validation capabilities.
+//
+// ref: https://github.com/kubernetes-sigs/controller-tools/blob/v0.19.0/pkg/crd/markers/
+func DefaultMarkerRules() map[string]MarkerScopeRule {
```
Does this need to be exported?

JoelSpeed commented on pkg/analysis/markerscope/analyzer.go:
```diff
@@ -0,0 +1,491 @@
+
+	return &analysis.Analyzer{
+		Name:             name,
+		Doc:              "Validates that markers are applied in the correct scope.",
```
I'd prefer to see more detailed documentation here please

JoelSpeed commented on pkg/analysis/markerscope/analyzer.go:
```diff
@@ -0,0 +1,491 @@
+		case *ast.GenDecl:
+			a.checkTypeMarkers(pass, node, markersAccess)
+		}
+	})
```
We have our own inspector helper that we prefer to use rather than this built in one, please check other linters and copy that pattern

JoelSpeed reviewed (commented):
Some of this code appears to be a little over complex for what we were expecting here. Lets try and get this cleaned up and simplified a bit.

Lets focus initially on what the public configuration should look like, and then look at the implementation I think

Out of interest, were any code assistants used to help prepare this PR? (I'm currently comparing different models so good to have context on what models are being used where)

sivchari commented on docs/linters.md:
```diff
@@ -289,6 +290,158 @@ lintersConfig:
+    allowDangerousTypes: false # Allow dangerous number types (float32, float64). Defaults to `false`.
+    markerRules:
+      # Override default rule for a built-in marker
+      "optional":
```
```suggestion
      overrideMarkers:
        - identifier: "optional":
```

I want override section to be explicit.

sivchari commented on docs/linters.md:
```diff
@@ -289,6 +290,158 @@ lintersConfig:
+        scope: field  # or: type, any
+
+      # Add a custom marker with scope constraint only
+      "mycompany:validation:CustomMarker":
```
```suggestion
     customMarkers:
      - identifier: "mycompany:validation:CustomMarker":
```

Same here and I think each marker should be defined as value, then we should have some key like `identifier`

sivchari commented on docs/linters.md:
```diff
@@ -289,6 +290,158 @@ lintersConfig:
+**Type constraint fields:**
+- `allowedSchemaTypes`: List of allowed OpenAPI schema types (`integer`, `number`, `string`, `boolean`, `array`, `object`)
+- `elementConstraint`: Nested constraint for array element types (only valid when `allowedSchemaTypes` includes `array`)
+- `strictTypeConstraint`: When `true`, markers with `AnyScope` and type constraints applied to fields using named types must be declared on the type definition instead of the field. Defaults to `false`.
```
Let's move this in front of YAML configuration

sivchari commented on pkg/analysis/markerscope/initializer.go:
```diff
@@ -0,0 +1,123 @@
+	return initializer.NewConfigurableInitializer(
+		name,
+		initAnalyzer,
+		false, // Not enabled by default
```
Why do you disable this by default ? Even if we don't have configuration for markerscope, the analysis result of this is beneficial for user since wrong markers are diagnosed.

nayuta723 replied:
At the initial release, it was unstable, so we set the default to false. 
However, since we believe this feature is beneficial for all users, we will change it to true.

sivchari commented on pkg/analysis/markerscope/schema.go:
```diff
@@ -0,0 +1,125 @@
+	}
+
+	return t
+}
```
Let's move this to utils package. We can reuse this in the future.

sivchari commented on pkg/analysis/markerscope/analyzer.go:
```diff
@@ -0,0 +1,491 @@
+}
+
+// extractIdent extracts an *ast.Ident from an ast.Expr, unwrapping pointers and arrays.
+func extractIdent(expr ast.Expr) *ast.Ident {
```
I think it's good to move to utils package.

nayuta723 committed:
refactor(markerscope): convert config to PascalCase values and named object lists

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 commented on pkg/analysis/markerscope/analyzer.go:
```diff
@@ -0,0 +1,491 @@
+		// Get the underlying type for better error messages
+		underlyingType := getUnderlyingType(t)
+		return &DengerousTypeError{Type: underlyingType.String()}
+	}
```
When DangerousTypes is enabled, it allows floats configuration. However, since there is a nofloat linter, this is no longer necessary and should be removed. 
https://github.com/kubernetes-sigs/kube-api-linter/pull/161/files#r2470349926

The original intention was to produce errors for patterns where SchemaTypeNumber (which includes float32, float64) is used when Float is not allowed (DangerousTypes disabled).

nayuta723 committed:
refactor(markerscope): separate overrideMarkers and customMarkers with validation

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
feat(markerscope): add default config and enhance documentation

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
refactor(markerscope): remove support for float/number types

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 commented on pkg/analysis/markerscope/errors.go:
I created this to resolve the go-err113 lint error.
ref: https://github.com/Djarvur/go-err113

nayuta723 committed:
refactor(markerscope): replace StrictTypeConstraint with NamedTypeConstraint and improve code quality

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
refactor(markers): remove Kubebuilder prefix from SSA topology marker constants

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 commented on pkg/analysis/markerscope/analyzer.go:
```diff
@@ -0,0 +1,491 @@
+						},
+					},
+				}
+			}
```
It seems like we can't use `continue`.

nayuta723 committed:
refactor(markerscope): improve error handling and code organization

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 commented on pkg/analysis/markerscope/analyzer.go:
```diff
@@ -0,0 +1,491 @@
+func (a *analyzer) checkFieldMarkers(pass *analysis.Pass, field *ast.Field, markersAccess markershelper.Markers) {
+	fieldMarkers := markersAccess.FieldMarkers(field)
+
+	for _, marker := range fieldMarkers.UnsortedList() {
```
Since it was unclear whether it should be included in utils, I've only implemented it as a function for now.
https://github.com/kubernetes-sigs/kube-api-linter/pull/161/commits/9c0a9b6d67c4f0145e8f7b8dd47acfb20e7b1e6f

nayuta723 commented on docs/linters.md:
```diff
@@ -289,6 +290,158 @@ lintersConfig:
+      "mycompany:validation:NumericLimit":
+        scope: any
+        strictTypeConstraint: true # Require declaration on type definition for named types
+        typeConstraint:
```
Is the following sufficient?
[6270f45](https://github.com/kubernetes-sigs/kube-api-linter/pull/161/commits/6270f455a313c49d1f16d7ab13673d1403a4130a)

sivchari replied:
Sounds good.

nayuta723 commented on pkg/analysis/markerscope/analyzer.go:
```diff
@@ -0,0 +1,491 @@
+		}
+	} else {
+		message = fmt.Sprintf("marker %q cannot be applied to types", marker.Identifier)
+	}
```
I have taken care of the matter below.
[9c0a9b6](https://github.com/kubernetes-sigs/kube-api-linter/pull/161/commits/9c0a9b6d67c4f0145e8f7b8dd47acfb20e7b1e6f)

nayuta723 committed:
refactor(markerscope): simplify getBasicTypeSchema with nolint:exhaustive

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 commented on pkg/analysis/markerscope/schema.go:
```diff
@@ -0,0 +1,125 @@
+		// These types are not supported in OpenAPI schemas
+		return ""
+	default:
+		return ""
```
I have taken care of the matter below.
[f969d45](https://github.com/kubernetes-sigs/kube-api-linter/pull/161/commits/f969d4536d8aaa2971bec9d0ae4aafccf70f9b20)

nayuta723 committed:
refactor(markerscope): update scope handling to support multiple scopes for markers

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

JoelSpeed commented on docs/linters.md:
```diff
@@ -383,6 +384,168 @@ lintersConfig:
+        scopes: [Field]  # Can specify [Field], [Type], or [Field, Type]
+
+    # Add rules for custom markers
+    customMarkers:
```
Do we need separate lists here? Or shall we just assume anything in the custom list that matches a built-in config overrides it?

JoelSpeed commented on pkg/analysis/markerscope/analyzer.go:
```diff
@@ -0,0 +1,492 @@
+	// analyzer needs to know about these markers
+	for marker := range a.markerRules {
+		markershelper.DefaultRegistry().Register(marker)
+	}
```
This is better here, I don't think we need the `init`

JoelSpeed commented on pkg/analysis/markerscope/analyzer.go:
```diff
@@ -0,0 +1,492 @@
+		// Check if FieldScope is allowed
+		if !rule.AllowsScope(FieldScope) {
+			a.reportFieldScopeViolation(pass, field, marker, rule)
+			continue
```
Why continue? Why not also check the type constraint?

JoelSpeed commented on pkg/analysis/markerscope/analyzer.go:
```diff
@@ -0,0 +1,492 @@
+}
+
+// reportFieldScopeViolation reports a scope violation for a field marker.
+func (a *analyzer) reportFieldScopeViolation(pass *analysis.Pass, field *ast.Field, marker markershelper.Marker, rule MarkerScopeRule) {
```
What if instead of reporting and doing logic in here, we had a function roughly like this that could be called by both the field and type side of things

```
checkAllowedScope(pass *analysis.Pass, node ast.Node, marker markershelper.Marker, appliedScope ScopeConstraint, rule MarkerScopeRule) {
  if rule.AllowsScope(appliedScope) {
    // Do nothing, the applied scope is allowed
    return
  }

  // Do reporting and suggesting fix work
  message = fmt.Sprintf("marker %s cannot be applied to scope %s, must be applied to scope(s): %s", marker.Identifier, appliedScope, strings.Join(rule.Scopes, ","))
}
```

JoelSpeed commented on pkg/analysis/markerscope/analyzer.go:
```diff
@@ -0,0 +1,492 @@
+	tv, ok := pass.TypesInfo.Types[field.Type]
+	if !ok {
+		return nil // Skip if we can't determine the type
+	}
```
What about using the utils LookupTypeSpec?

JoelSpeed commented on pkg/analysis/markerscope/initializer.go:
```diff
@@ -0,0 +1,214 @@
+}
+
+func validatePolicy(policy MarkerScopePolicy, fldPath *field.Path) field.ErrorList {
+	if policy != "" && policy != MarkerScopePolicyWarn && policy != MarkerScopePolicySuggestFix {
```
Make this a switch statement, it will be neater

JoelSpeed commented on pkg/analysis/markerscope/marker_rules.go:
```diff
@@ -0,0 +1,494 @@
+}
+
+// addGeneralMarkers adds general markers that can apply to any type.
+func addGeneralMarkers(rules map[string]MarkerScopeRule) {
```
Why are these not with the FieldOrType markers?

JoelSpeed commented on pkg/analysis/markerscope/initializer_test.go:
```diff
@@ -0,0 +1,350 @@
+			Entry("With empty config", testCase{
+				config:      markerscope.MarkerScopeConfig{},
+				expectedErr: "",
+			}),
```
These two are the same?

JoelSpeed commented on pkg/analysis/markerscope/initializer_test.go:
```diff
@@ -0,0 +1,350 @@
+				expectedErr: "",
+			}),
+
+			Entry("With valid suggest_fix policy", testCase{
```
Why is it using snake case here?
```suggestion
			Entry("With valid SuggestFix policy", testCase{
```

JoelSpeed commented on pkg/analysis/markerscope/initializer_test.go:
```diff
@@ -0,0 +1,350 @@
+						},
+					},
+				},
+				expectedErr: `scope is required`,
```
Can you demonstrate the full error including the field path? Or do we not get that here?

JoelSpeed commented on pkg/analysis/markerscope/initializer_test.go:
```diff
@@ -0,0 +1,350 @@
+				expectedErr: "",
+			}),
+
+			Entry("With valid type constraint with multiple types", testCase{
```
Description doesn't match the test case. It's only checking one allowed schema type and is the same as the case above

JoelSpeed commented on pkg/analysis/markerscope/initializer_test.go:
```diff
@@ -0,0 +1,350 @@
+	Context("analyzer initialization", func() {
+		It("should initialize analyzer with nil config", func() {
+			// Note: Init expects a MarkerScopeConfig, passing nil will error
+			// Use empty config instead
```
So why not check the error?

nayuta723 replied:
This is missing. The document is incorrect, so I will correct it.

JoelSpeed reviewed (commented):
Apologies for the delay in reviewing this, it's a rather large and complex PR and it's taken me quite some time today to review (I haven't actually reviewed the test cases even yet).

I'd like you to review my comments and lets focus on trying to get the docs/config we want nailed, and then simplifying the implementation by sharing some more of the code between the type and field paths where possible

nayuta723 committed:
refactor(markerscope): rename and update NamedTypeConstraint values for clarity

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
refactor(markerscope): update default policy and improve documentation improved readability.

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
refactor(markerscope): update scope definitions and enhance error messages for clarity

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
refactor(markerscope): consolidate marker validation logic and enhance error reporting

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
refactor(utils): enhance type unwrapping and extraction to include map types

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
refactor(markerscope): remove general markers and enhance field-or-type marker definitions for clarity

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
refactor(markerscope): improve code organization and enhance schema type validation

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
refactor(markerscope): streamline type constraint validation and enhance error reporting

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
docs(linters): add mapping for OpenAPI 'number' type to Go types

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
refactor(markerscope)!: merge OverrideMarkers into CustomMarkers

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
feat(markerscope): add support for number type in numeric markers

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
refactor(markerscope): replace ambiguous "Any" terminology with explicit descriptions

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
feat(registration): register markerscope linter

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
fix(markerscope): treat []byte as string type for OpenAPI compatibility

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

sivchari commented on pkg/markers/markers.go:
```diff
@@ -153,14 +153,38 @@ const (
+	MapTypeMarker = "mapType"
+
+	// StructTypeMarker is the marker used to specify the atomicity level of a struct.
+	StructTypeMarker = "structType"
```
Let's move these markers to after L: 29

sivchari commented on pkg/analysis/utils/utils.go:
```diff
@@ -549,3 +549,92 @@ func getFieldTypeName(field *ast.Field) string {
+// For slice and array types, it returns the element type.
+// For map types, it returns the value type.
+// Otherwise, it returns the type as-is.
+func UnwrapType(t types.Type) types.Type {
```
```suggestion
func Underlying(t types.Type) types.Type {
```

nayuta723 committed:
refactor(markerscope): consolidate type constraint validation

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
test(markerscope): add type override validation tests

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
docs(markerscope): fix lint errors and document Type override feature

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 commented:
@JoelSpeed 
I’ve updated the configuration based on the following comments:
- https://github.com/kubernetes-sigs/kube-api-linter/pull/161#discussion_r2602258878
- https://github.com/kubernetes-sigs/kube-api-linter/pull/161#discussion_r2602598602

Also, I’ve addressed your feedback. Please take a look.

JoelSpeed commented on docs/linters.md:
```diff
@@ -424,6 +425,198 @@ lintersConfig:
+    # Custom markers can be used for both:
+    # 1. Overriding built-in markers (by using the same identifier)
+    # 2. Adding validation for your own custom markers
+    customMarkers:
```
Do we need to include `custom` here? What do we do in other places? Perhaps just `markers` would be sufficient?

JoelSpeed commented on pkg/analysis/markerscope/analyzer.go:
```diff
@@ -0,0 +1,560 @@
+
+func (a *analyzer) extractMarkerText(marker markershelper.Marker) string {
+	return marker.RawComment + "\n"
+}
```
Why is this a method on the analyzer?

Extract seems wrong here too, maybe `rawMarkerLine`?

JoelSpeed commented on pkg/analysis/markerscope/analyzer.go:
```diff
@@ -0,0 +1,560 @@
+	rule MarkerScopeRule,
+) {
+	// Determine the alternate scope
+	alternateScope := appliedScope.Opposite()
```
I'm not sure about this opposite concept, I'd rather something explicit here like a switch statement. Opposite implies there can only ever be two states, and I'd rather we don't bake that assumption in (e.g. what if we want to support package level markers in the future?)

JoelSpeed commented on pkg/analysis/markerscope/config.go:
```diff
@@ -0,0 +1,138 @@
+	// namedTypeConstraint specifies how markers should be applied to named types.
+	// When a field uses a named type (e.g., type CustomInt int32), this determines
+	// whether the marker can be on the field or must be on the type definition.
+	// Valid values: AllowTypeOrField, OnTypeOnly.
```
Is there a use case for `OnFieldOnly`?

JoelSpeed commented on pkg/analysis/markerscope/initializer.go:
```diff
@@ -0,0 +1,166 @@
+
+func isValidNamedTypeConstraint(ntc NamedTypeConstraint) bool {
+	switch ntc {
+	case "", NamedTypeConstraintAllowTypeOrField, NamedTypeConstraintOnTypeOnly:
```
What happens if this is configured as OnTypeOnly, but the valid scope is only `Field`?

JoelSpeed commented on pkg/analysis/markerscope/analyzer_test.go:
```diff
@@ -0,0 +1,121 @@
+func TestAnalyzerWithDefaultConfig(t *testing.T) {
+	testdata := analysistest.TestData()
+	// Test with nil config - should use all defaults:
+	// - Policy: Warn
```
SuggestFix is the default?

JoelSpeed commented on pkg/analysis/markerscope/testdata/src/a/array.go:
```diff
@@ -0,0 +1,99 @@
+	ValidBooleanArrayType BooleanArray `json:"validBooleanArrayType"`
+
+	// Invalid: Field marker on named array type (should use type definition)
+	// +kubebuilder:validation:MinItems=5 // want `marker "kubebuilder:validation:MinItems": marker should be declared on the type definition of FieldMarkerOnNamedArray instead of the field`
```
If this is unhelpful, how easy is it to reconfigure this?

JoelSpeed commented on pkg/analysis/markerscope/testdata/src/a/items.go.golden:
```diff
@@ -0,0 +1,147 @@
+type InvalidItemsPatternOnIntArrayType []int32
+
+// Invalid: items:MinProperties on string array type
+type InvalidItemsMinPropertiesOnStringArrayType []string
```
Shouldn't the markers have moved to here? They've been removed from the original location?

everettraven reviewed (commented):
A couple high-level comments.

At a brief glance, it looks like Joel has done a pretty thorough review and I agree with most, if not all, of his comments.

I'll circle back around for a more thorough review some time this week.

nayuta723 committed:
docs: update Type Override examples to real-world use cases

Replace artificial examples with practical ones based on json.RawMessage
and metav1.Duration patterns. Add invalid test cases for completeness.

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 commented on docs/linters.md:
```diff
@@ -424,6 +425,198 @@ lintersConfig:
+    customMarkers:
+      # Override a built-in marker by using its identifier
+      - identifier: "optional"
+        scopes: [Field, Type]  # Override default [Field] to allow on both
```
That makes sense. 
So We should avoid modifying defaultMarker and only define the behavior for the new types, right?

nayuta723 committed:
refactor(markerscope): use explicit scope handling instead of Opposite()

Avoid baking in the assumption that only two scopes exist by using
explicit switch statements and rule.Scopes directly.

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
refactor(markerscope): replace extractMarkerText with RawMarkerLine utility

Removed the extractMarkerText method and replaced its usage with the new RawMarkerLine function for better clarity and consistency in handling marker comments.

Signed-off-by: nayuta-ai <nayuta723@gmail.com>

nayuta723 committed:
test(markerscope): update analyzer tests and fix marker comments

Signed-off-by: nayuta723 <nayuta723@gmail.com>

nayuta723 committed:
test(markerscope): add validation tests for named type constraints

Signed-off-by: nayuta723 <nayuta723@gmail.com>

nayuta723 committed:
test(markerscope): add comprehensive tests for FieldOrType markers

Signed-off-by: nayuta723 <nayuta723@gmail.com>

sivchari reviewed (commented):
Thank you for working on it! I left nits.

nayuta723 committed:
refactor(markerscope): improve named type constraint validation

Updated the validateNamedTypeConstraint function to return field errors instead of a single error, enhancing error reporting for invalid constraints. Simplified the logic for checking TypeScope inclusion using slices.Contains. Removed unnecessary error handling in the analyzer reportScopeViolation function.

Signed-off-by: nayuta723 <nayuta723@gmail.com>

nayuta723 committed:
docs(linters): update named type constraints to named type preferences

Revised documentation to clarify the `namedTypePreference` field, including its valid values and precedence rules. Enhanced examples to demonstrate global and per-marker settings for uniform behavior across markers. This change aims to improve understanding of marker placement in relation to type and field scopes.

Signed-off-by: nayuta723 <nayuta723@gmail.com>

nayuta723 committed:
feat(markers): add new markers for list and map types

Introduced new markers: ListTypeMarker, ListMapKeyMarker, MapTypeMarker, and StructTypeMarker to enhance server-side apply operations and atomicity specifications. Removed redundant marker definitions to streamline the codebase.

Signed-off-by: nayuta723 <nayuta723@gmail.com>

nayuta723 committed:
refactor(markerscope): streamline marker sorting and rename utility function

Removed the sortMarkersByPosition function and replaced its usage with the new SortedMarkers utility in the analyzer. Renamed UnwrapType to Underlying for clarity in the utils package, ensuring consistent terminology for type handling.

Signed-off-by: nayuta723 <nayuta723@gmail.com>

nayuta723 committed:
refactor(markerscope): update inspector helper import and usage

Replaced the inspectorhelper import with the inspector import in the analyzer. Updated the references in the analyzer's run method to align with the new import, ensuring consistent usage of the inspector functionality.

Signed-off-by: nayuta723 <nayuta723@gmail.com>

nayuta723 committed:
refactor(markerscope): enhance type definition validation logic

Updated the handleShouldBeOnTypeDefinition function to improve validation for marker placement. The logic now allows for more flexible constraints by skipping validation if the marker permits placement on both fields and types, or if TypeScope is not allowed.

Signed-off-by: nayuta723 <nayuta723@gmail.com>

nayuta723 committed:
docs(markerscope): enhance documentation for marker scope validation and named type preferences

Updated the documentation to clarify the rules for marker placement on fields and types, including the introduction of the namedTypePreference setting. Added examples to illustrate valid configurations and the implications of using OnTypeOnly for marker declarations.

Signed-off-by: nayuta723 <nayuta723@gmail.com>

nayuta723 committed:
refactor(markerscope): optimize marker validation with type usage caching

Introduced a type usage cache to enhance the efficiency of marker validation by enabling O(1) lookups for fields associated with type specifications. Updated the analyzer's methods to utilize this cache, reducing redundant traversals during marker checks. Removed the deprecated LookupTypeSpecUsage function from the utils package to streamline the codebase.

Signed-off-by: nayuta723 <nayuta723@gmail.com>
